{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Prepare MitoFish data\n",
    "\n",
    "MitoFish website: http://mitofish.aori.u-tokyo.ac.jp/download.html\n",
    "\n",
    "##### Download and process mitoannotator (complete mitogenome) records:\n",
    "\n",
    "```\n",
    "# Download and unzip file in a folder named 'mitoannotations'\n",
    "mkdir mitoannotations\n",
    "cd mitoannotations\n",
    "wget http://mitofish.aori.u-tokyo.ac.jp/files/mitoannotations.zip\n",
    "unzip mitoannotations.zip\n",
    "\n",
    "# get accession numbers\n",
    "ls *.txt | cut -d '_' -f1,2 >complete.accession\n",
    "\n",
    "# get species names\n",
    "ls *.txt | cut -d '_' -f3- | sed \"s/.txt/#complete mitogenomes/g\" >complete.species\n",
    "\n",
    "# Make a list of accession number and species, separated by \"#\"\n",
    "paste -d \"#\" complete.accession complete.species >complete.list\n",
    "\n",
    "```\n",
    "##### Download and process complete+partial mDNA sequence file: \n",
    "\n",
    "```\n",
    "# Download and unzip sequence file in the same \"mitoannotations\" folder\n",
    "wget http://mitofish.aori.u-tokyo.ac.jp/files/complete_partial_mitogenomes.zip\n",
    "unzip complete_partial_mitogenomes.zip\n",
    "\n",
    "# get accession numbers\n",
    "grep \">\" complete_partial_mitogenomes.fa | awk -F \"|\" '{print $2}' >mitofish.accession\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Prepare NCBI data\n",
    "\n",
    "NCBI blast databases ftp site: https://ftp.ncbi.nlm.nih.gov/blast/db/FASTA/\n",
    "\n",
    "```\n",
    "# Download and unzip NCBI sequence file in the \"NCBI\" subfolder\n",
    "mkdir NCBI\n",
    "cd NCBI\n",
    "wget https://ftp.ncbi.nlm.nih.gov/blast/db/FASTA/nt.gz\n",
    "wget https://ftp.ncbi.nlm.nih.gov/blast/db/FASTA/nt.gz.md5\n",
    "md5sum -c nt.gz.md5 >md5sum.log\n",
    "unzip nt.gz\n",
    "\n",
    "# get accession numbers\n",
    "grep \">\" nt | cut -d ' ' -f1 | tr -d \">\"  >nt.accession \n",
    "\n",
    "# get gene names\n",
    "grep \">\" nt | cut -d ' ' -f2- >nt.genenames \n",
    "\n",
    "# Make a list of accession number and species, separated by \"#\"\n",
    "paste -d \"#\" nt.accession nt.genenames >nt.list\n",
    "cd ..\n",
    "```\n",
    "\n",
    "#### Extract (grep) 12S genes from NCBI records\n",
    "\n",
    "```\n",
    "grep -e \"12S ribosomal\" -e \"12S rRNA\" nt.list >12S.list\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This python script creates a dictionary from the nt.list file\n",
    "# The key is the NCBI accession number and the value is the gene description\n",
    "# This dictionary is then saved as a pickle named nt.pickle\n",
    "\n",
    "#!/usr/bin/python\n",
    "import pickle\n",
    "ntdict = {'Accession':'Gene description'}\n",
    "\n",
    "with open(\"nt.list\",'r') as f:\n",
    "    for line in f:\n",
    "        line = line.rstrip()\n",
    "        entry = line.split(\"#\")\n",
    "        fullacc=entry[0].split(\".\")\n",
    "        newentry={fullacc[0]:entry[1]} \n",
    "        ntdict.update(newentry)\n",
    "\n",
    "with open(\"nt.pickle\", 'wb') as handle:\n",
    "    pickle.dump(ntdict, handle, protocol=pickle.HIGHEST_PROTOCOL)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==== Searching... ====\n",
      "SVSP109#No hit found!\n",
      "alpha-lactalbumin#No hit found!\n",
      "LN610210#No hit found!\n",
      "LN610216#No hit found!\n",
      "miami#No hit found!\n",
      "==== Run complete! ===\n",
      "Total: 11 accession numbers\n",
      "No hits for 5 input accession numbers!\n"
     ]
    }
   ],
   "source": [
    "# This python script loads nt.pickle \n",
    "# Then, it matches MitoFish entries with keys (NCBI accession numbers) in the pickle\n",
    "# It takes ~24 minutes to match 616,999 MitoFish records to 57,377,397 NCBI GenBank records\n",
    "# cpupercent=101,cput=00:22:53,mem=34500088kb,ncpus=24,vmem=34845840kb,walltime=00:23:34\n",
    "\n",
    "#!/usr/bin/python\n",
    "\n",
    "import pickle\n",
    "\n",
    "import sys\n",
    "import os.path\n",
    "from os import path\n",
    "\n",
    "with open(\"nt.pickle\", 'rb') as handle:\n",
    "    NCBI = pickle.load(handle)\n",
    "\n",
    "outfile=\"mitofish.genes\"\n",
    "nohit=0\n",
    "count=0\n",
    "length=(len(NCBI))\n",
    "\n",
    "if path.exists(outfile) :\n",
    "    sys.exit(\"Error: Output file exists! Please rename output file and try again!\")\n",
    "\n",
    "output=open(outfile,'a')\n",
    "\n",
    "print(\"==== Searching... ====\")\n",
    "\n",
    "with open(\"mitofish.accession\",'r') as infile:\n",
    "    for inline in infile:\n",
    "        inline = inline.rstrip()\n",
    "        count +=1\n",
    "        if inline in NCBI:\n",
    "            output.write(\"%s#%s\\n\" % (inline,NCBI[inline]))\n",
    "        elif inline not in NCBI:\n",
    "            output.write(\"%s#No hit found!\\n\" % inline)\n",
    "            print(\"%s#No hit found!\" % inline)\n",
    "            nohit +=1\n",
    "            \n",
    "output.close()\n",
    "\n",
    "print (\"==== Run complete! ===\")\n",
    "print (\"Total: %d accession numbers\" % count)\n",
    "print (\"No hits for %d input accession numbers!\" % nohit)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pick up \"missing\" accession numbers \n",
    "\n",
    "In NCBI fasta files, records with duplicate sequences will be concatenated in the header line, e.g.\n",
    "\n",
    "```\n",
    ">LN610233.1 Chiloglanis anoterus mitochondrial partial D-loop, specimen voucher 68321_34LN610234.1 Chiloglanis anoterus mitochondrial partial D-loop, specimen voucher 68330_55\n",
    "```\n",
    "These will not be picked up by the python script above. First, let's extract the accession numbers with no hits into an output file named <b>nohit.accession</b>.\n",
    "```\n",
    "grep \"No hit\" mitofish.genes | awk -F \"#\" '{print $1}' | sed \"s/$/.[0-9]/g\" >nohit.accession\n",
    "```\n",
    "We added some regular expression patterns to the accession numbers with no hits so that the grep search is more specific. The output file <b>nohit.accession</b> looks like this:\n",
    "\n",
    "```\n",
    "LN610210.[0-9]\n",
    "LN610216.[0-9]\n",
    "LN610224.[0-9]\n",
    "LN610229.[0-9]\n",
    "LN610230.[0-9]\n",
    "LN610231.[0-9]\n",
    "LN610232.[0-9]\n",
    "LN610234.[0-9]\n",
    "LN610235.[0-9]\n",
    "```\n",
    "We will also extract the hits from the python search for later use. Output file is <b>hit.list</b>:\n",
    "```\n",
    "grep -v \"No hit\" mitofish.genes >hit.list\n",
    "```\n",
    "\n",
    "Next, split up the reference file (<b>nt.genenames</b> generated in previous step) into smaller chunks of 1 million lines each. Each split file will have the prefix x followed by a number e.g. x00, x01\n",
    "\n",
    "```\n",
    "cd NCBI\n",
    "split -d -l 1000000 nt.genenames\n",
    "cd ..\n",
    "```\n",
    "\n",
    "For each accession number with no hit, search each chunk of nt.genenames for matches using all processors (-P 0). -n 1 specifies the number of argument to pass per command line. This takes ~39 hours to run for 144,308 accession numbers.\n",
    "\n",
    "```\n",
    "for id in $(cat nohit.accession)\n",
    "do\n",
    "string=`ls NCBI/x* | xargs -n 1 -P 0 grep $id`\n",
    "echo \"$id%$string\" >>nohit.genes\n",
    "done\n",
    "```\n",
    "Output file is <b>nohit.genes</b>. First, let's clean up the output file by removing non-specific matches and removing the regular expression patterns following the accession numbers:\n",
    "\n",
    "```\n",
    "grep \"#\" nohit.genes | sed \"s/.\\[0\\-9\\]//g\" >nohit.genes.clean\n",
    "```\n",
    "\n",
    "Now, combine <b>hit.list</b> (accession numbers with exact matches with NCBI accession numbers) with <b>nohit.genes.clean</b> (accession numbers with duplicated sequences). The output file will be <b>mitofish.hit.list</b>:\n",
    "```\n",
    "cat hit.list nohit.genes.clean >mitofish.hit.list\n",
    "```\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "subspecies\n",
      "test\n",
      "12S.list\n",
      "=== Searching query #1 ===\n",
      "Histioteuthis celetaria celetaria hits (full taxonomy): 0\n",
      "Histioteuthis celetaria hits (species): 0\n",
      "Histioteuthis hits (genus): 1\n",
      "=== Searching query #2 ===\n",
      "Histioteuthis corona corona hits (full taxonomy): 0\n",
      "Histioteuthis corona hits (species): 0\n",
      "Histioteuthis hits (genus): 1\n",
      "=== Searching query #3 ===\n",
      "Lampadena urophaos atlantica hits (full taxonomy): 0\n",
      "Lampadena urophaos hits (species): 3\n",
      "Lampadena hits (genus): 5\n",
      "=== Searching query #4 ===\n",
      "Notoscopelus elongatus kroyeri hits (full taxonomy): 0\n",
      "Notoscopelus elongatus hits (species): 0\n",
      "Notoscopelus hits (genus): 7\n",
      "=== Searching query #5 ===\n",
      "Scopelogadus mizolepis mizolepis hits (full taxonomy): 0\n",
      "Scopelogadus mizolepis hits (species): 0\n",
      "Scopelogadus hits (genus): 5\n",
      "=== Searching query #6 ===\n",
      "Stomias boa boa hits (full taxonomy): 0\n",
      "Stomias boa hits (species): 2\n",
      "Stomias hits (genus): 11\n",
      "==== Run complete! ===\n",
      "Accession number of full taxonomy hits and description saved in <test_fulltaxonomy.hits.txt>\n",
      "Accession number of species hits and description saved in <test_species.hits.txt>\n",
      "Accession number of genus hits and description saved in <test_genus.hits.txt>\n"
     ]
    }
   ],
   "source": [
    "# From a user-provided list of genera/species/subspecies, this script extracts the corresponding GenBank accession numbers \n",
    "# and gene names of their 12S rRNA sequences or mitochondrial sequences, if available.\n",
    "\n",
    "# This script is interactive and takes 3 inputs, in the following order:\n",
    "# 1) Input file name with extension (e.g. input.txt)\n",
    "# 2) Output file prefix for output files: <prefix>_genus.hits.txt, <prefix>_species.hits.txt, <prefix>_fulltaxonomy.hits.txt\n",
    "# 3) Reference database - either 12S.list or mitofish.hit.list\n",
    "# Output files are separated by \"#\" in the format: GenBank_Accession#Gene_description\n",
    "\n",
    "\n",
    "import sys\n",
    "import os.path\n",
    "from os import path\n",
    "\n",
    "# Modify input file name here\n",
    "input_file = input()\n",
    "ref=tuple(open(input_file,'r'))\n",
    "\n",
    "\n",
    "output_prefix = input()\n",
    "\n",
    "full_path=str(output_prefix+\"_fulltaxonomy.hits.txt\")\n",
    "species_path=str(output_prefix+\"_species.hits.txt\")\n",
    "genus_path=str(output_prefix+\"_genus.hits.txt\")\n",
    "\n",
    "# Throw an error message and exit if output file(s) already exist\n",
    "\n",
    "if path.exists(full_path) or path.exists(species_path) or path.exists(genus_path) :\n",
    "    sys.exit(\"Error: Output file exists! Please rename output file and try again!\")\n",
    "\n",
    "# Create output files if all is well\n",
    "                  \n",
    "full_hit=open(full_path,'a')\n",
    "species_hit = open(species_path,'a')\n",
    "genus_hit = open(genus_path,'a')\n",
    "\n",
    "# Create sets to hold lines already seen, for line deduplication later\n",
    "full_seen=set()\n",
    "species_seen=set()\n",
    "genus_seen=set()\n",
    "\n",
    "reference_file = input()\n",
    "    \n",
    "i = 0\n",
    "while (i < len(ref)):\n",
    "    \n",
    "    taxa=str(ref[i]).rsplit()\n",
    "    fulltaxa=str(ref[i]).rsplit(\"\\n\")\n",
    "    fullquery=str(fulltaxa[0])\n",
    "    squery=str(taxa[0]+\" \"+taxa[1])\n",
    "    gquery=str(taxa[0])\n",
    "    \n",
    "    fullcount=0\n",
    "    scount=0\n",
    "    gcount=0\n",
    "    \n",
    "    qcount = i+1\n",
    "    print (\"=== Searching query #%d ===\" % qcount)\n",
    "    \n",
    "    with open(reference_file, 'r') as f:\n",
    "        for line in f.readlines():\n",
    "            if fullquery in line:\n",
    "                fullcount += 1\n",
    "                if line not in full_seen:\n",
    "                     full_hit.write(line)\n",
    "                full_seen.add(line)\n",
    "            if squery in line:\n",
    "                scount += 1\n",
    "                if line not in species_seen:\n",
    "                     species_hit.write(line)\n",
    "                species_seen.add(line)\n",
    "            if gquery in line:\n",
    "                gcount += 1 \n",
    "                if line not in genus_seen:\n",
    "                     genus_hit.write(line)\n",
    "                genus_seen.add(line)\n",
    "\n",
    "    print(\"%s hits (full taxonomy): %d\" % (fullquery,fullcount))\n",
    "    print(\"%s hits (species): %d\" % (squery,scount))\n",
    "    print(\"%s hits (genus): %d\" % (gquery,gcount))\n",
    "    \n",
    "    i += 1\n",
    "\n",
    "full_hit.close()\n",
    "species_hit.close()\n",
    "genus_hit.close()\n",
    "\n",
    "print (\"==== Run complete! ===\")\n",
    "print (\"Accession number of full taxonomy hits and description saved in <%s_fulltaxonomy.hits.txt>\" % output_prefix)\n",
    "print (\"Accession number of species hits and description saved in <%s_species.hits.txt>\" % output_prefix)\n",
    "print (\"Accession number of genus hits and description saved in <%s_genus.hits.txt>\" % output_prefix)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "BioPython",
   "language": "python",
   "name": "biopython"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
