{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Prepare KpH-2019703 catchsample list of species\n",
    "\n",
    "Remove non-standard characters `Å`, `Ø`, and `é` using SublimeText:\n",
    "\n",
    "```\n",
    "cd /Users/luke/NOAA/Cruises/KpH-May2019/fishdata\n",
    "cp catchsample.txt catchsample_stdchars.txt\n",
    "# remove non-standard characters in catchsample_stdchars.txt\n",
    "```\n",
    "\n",
    "Parse catchsample to get lists of unique taxa: subspecies, species, genera, families (-ae), orders (-es) from field \"scientificname\":\n",
    "\n",
    "```\n",
    "# subspecies (three or more words)\n",
    "cat catchsample_stdchars.txt | cut -d\"<TAB>\" -f11 | perl -lne '@F=split / /; if (@F>2) {print \"@F[0] @F[1] @F[2]\"}' | sort | uniq > catchsample_stdchars_subspecies_uniq.txt\n",
    "\n",
    "# species (two words)\n",
    "cat catchsample_stdchars.txt | cut -d\"<TAB>\" -f11 | perl -lne '@F=split / /; if (@F==2) {print \"@F[0] @F[1]\"}' | sort | uniq > catchsample_stdchars_species_uniq.txt\n",
    "\n",
    "# genera (one word NOT ending in -ae or -es)\n",
    "cat catchsample_stdchars.txt | cut -d\"<TAB>\" -f11 | perl -lne '@F=split / /; if (@F==1) {print @F[0]}' | egrep -v \"^[A-Z][a-z]*ae$\" | egrep -v \"^[A-Z][a-z]*es$\" | egrep -v \"scientificname\" | sort | uniq > catchsample_stdchars_genera_uniq.txt\n",
    "\n",
    "# familes (one word ending in -ae)\n",
    "cat catchsample_stdchars.txt | cut -d\"<TAB>\" -f11 | perl -lne '@F=split / /; if (@F==1) {print @F[0]}' | egrep \"^[A-Z][a-z]*ae$\" | sort | uniq > catchsample_stdchars_families_uniq.txt\n",
    "\n",
    "# orders (one word ending in -es)\n",
    "cat catchsample_stdchars.txt | cut -d\"<TAB>\" -f11 | perl -lne '@F=split / /; if (@F==1) {print @F[0]}' | egrep \"^[A-Z][a-z]*es$\" | sort | uniq > catchsample_stdchars_orders_uniq.txt\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Prepare MitoFish data\n",
    "\n",
    "MitoFish website: http://mitofish.aori.u-tokyo.ac.jp/download.html\n",
    "\n",
    "##### Download and process mitoannotator (complete mitogenome) records:\n",
    "\n",
    "```\n",
    "# Download and unzip file in a folder named 'mitoannotations'\n",
    "mkdir mitoannotations\n",
    "cd mitoannotations\n",
    "wget http://mitofish.aori.u-tokyo.ac.jp/files/mitoannotations.zip\n",
    "unzip mitoannotations.zip\n",
    "\n",
    "# get accession numbers\n",
    "ls *.txt | cut -d '_' -f1,2 >complete.accession\n",
    "\n",
    "# get species names\n",
    "ls *.txt | cut -d '_' -f3- | sed \"s/.txt/#complete mitogenomes/g\" >complete.species\n",
    "\n",
    "# Make a list of accession number and species, separated by \"#\"\n",
    "paste -d \"#\" complete.accession complete.species >complete.list\n",
    "\n",
    "```\n",
    "##### Download and process complete+partial mDNA sequence file: \n",
    "\n",
    "```\n",
    "# Download and unzip sequence file in the same \"mitoannotations\" folder\n",
    "wget http://mitofish.aori.u-tokyo.ac.jp/files/complete_partial_mitogenomes.zip\n",
    "unzip complete_partial_mitogenomes.zip\n",
    "\n",
    "# get accession numbers\n",
    "grep \">\" complete_partial_mitogenomes.fa | awk -F \"|\" '{print $2}' >mitofish.accession\n",
    "\n",
    "# get species names\n",
    "grep \">\" complete_partial_mitogenomes.fa | awk -F \"|\" '{print $3}' | sed \"s/ (\\[.*\\])$//g\" | tr \" \" \"_\" >mitofish.species\n",
    "\n",
    "# Make a list of accession number and species, separated by \"#\"\n",
    "paste -d \"#\" mitofish.accession mitofish.species >mitofish.list\n",
    "\n",
    "```\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Prepare NCBI data\n",
    "\n",
    "NCBI blast databases ftp site: https://ftp.ncbi.nlm.nih.gov/blast/db/FASTA/\n",
    "\n",
    "```\n",
    "# Download and unzip NCBI sequence file in the \"NCBI\" subfolder\n",
    "mkdir NCBI\n",
    "cd NCBI\n",
    "wget https://ftp.ncbi.nlm.nih.gov/blast/db/FASTA/nt.gz\n",
    "wget https://ftp.ncbi.nlm.nih.gov/blast/db/FASTA/nt.gz.md5\n",
    "md5sum -c nt.gz.md5 >md5sum.log\n",
    "unzip nt.gz\n",
    "\n",
    "# get accession numbers\n",
    "grep \">\" nt | cut -d ' ' -f1 | tr -d \">\"  >nt.accession \n",
    "\n",
    "# get gene names\n",
    "grep \">\" nt | cut -d ' ' -f2- >nt.genenames \n",
    "\n",
    "# Make a list of accession number and species, separated by \"#\"\n",
    "paste -d \"#\" nt.accession nt.genenames >nt.list\n",
    "\n",
    "# Split the list into smaller files of 10 million rows each\n",
    "split -d -l 1000000 nt.list\n",
    "\n",
    "# Each split file will have the prefix x followed by a number e.g. x00, x01\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Extract (grep) NCBI gene names from MitoFish accession (excluding complete mitogenomes)\n",
    "```\n",
    "\n",
    "for id in $(cat ../mitofish.accession); do ls x* | xargs -n 1 -P 0 grep $id >>../mitofish.genenames ; done\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Extract (grep) 12S genes from NCBI records\n",
    "\n",
    "```\n",
    "grep -e \"12S ribosomal\" -e \"12S rRNA\" nt.list >12S.list\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=== Searching query #1 ===\n",
      "Histioteuthis celetaria celetaria hits (full taxonomy): 0\n",
      "Histioteuthis celetaria hits (species): 0\n",
      "Histioteuthis hits (genus): 1\n",
      "=== Searching query #2 ===\n",
      "Histioteuthis corona corona hits (full taxonomy): 0\n",
      "Histioteuthis corona hits (species): 0\n",
      "Histioteuthis hits (genus): 1\n",
      "=== Searching query #3 ===\n",
      "Lampadena urophaos atlantica hits (full taxonomy): 0\n",
      "Lampadena urophaos hits (species): 3\n",
      "Lampadena hits (genus): 5\n",
      "=== Searching query #4 ===\n",
      "Notoscopelus elongatus kroyeri hits (full taxonomy): 0\n",
      "Notoscopelus elongatus hits (species): 0\n",
      "Notoscopelus hits (genus): 7\n",
      "=== Searching query #5 ===\n",
      "Scopelogadus mizolepis mizolepis hits (full taxonomy): 0\n",
      "Scopelogadus mizolepis hits (species): 0\n",
      "Scopelogadus hits (genus): 5\n",
      "=== Searching query #6 ===\n",
      "Stomias boa boa hits (full taxonomy): 0\n",
      "Stomias boa hits (species): 2\n",
      "Stomias hits (genus): 11\n",
      "==== Run complete! ===\n",
      "Accession number of full taxonomy hits and description saved in <newtest_fulltaxonomy.hits.txt>\n",
      "Accession number of species hits and description saved in <newtest_species.hits.txt>\n",
      "Accession number of genus hits and description saved in <newtest_genus.hits.txt>\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "import os.path\n",
    "from os import path\n",
    "\n",
    "# Modify input file name here\n",
    "input_file = 'fishdata/catchsample_stdchars_subspecies_uniq.txt'\n",
    "ref=tuple(open(input_file,'r'))\n",
    "\n",
    "# Modify output file prefix here\n",
    "# 3 output files will be generated = <prefix>_genus.hits.txt, <prefix>_species.hits.txt and <prefix>_fulltaxonomy.hits.txt\n",
    "# Output file is separated by \"#\" in the format: GenBank_Accession#Gene_description\n",
    "\n",
    "output_prefix = 'newtest'\n",
    "\n",
    "full_path=str(output_prefix+\"_fulltaxonomy.hits.txt\")\n",
    "species_path=str(output_prefix+\"_species.hits.txt\")\n",
    "genus_path=str(output_prefix+\"_genus.hits.txt\")\n",
    "\n",
    "# Throw an error message and exit if output file(s) already exist\n",
    "\n",
    "if path.exists(full_path) or path.exists(species_path) or path.exists(genus_path) :\n",
    "    sys.exit(\"Error: Output file exists! Please rename output file and try again!\")\n",
    "\n",
    "# Create output files if all is well\n",
    "                  \n",
    "full_hit=open(full_path,'a')\n",
    "species_hit = open(species_path,'a')\n",
    "genus_hit = open(genus_path,'a')\n",
    "\n",
    "# Create sets to hold lines already seen, for line deduplication later\n",
    "full_seen=set()\n",
    "species_seen=set()\n",
    "genus_seen=set()\n",
    "    \n",
    "i = 0\n",
    "while (i < len(ref)):\n",
    "    \n",
    "    taxa=str(ref[i]).rsplit()\n",
    "    fulltaxa=str(ref[i]).rsplit(\"\\n\")\n",
    "    fullquery=str(fulltaxa[0])\n",
    "    squery=str(taxa[0]+\" \"+taxa[1])\n",
    "    gquery=str(taxa[0])\n",
    "    \n",
    "    fullcount=0\n",
    "    scount=0\n",
    "    gcount=0\n",
    "    \n",
    "    qcount = i+1\n",
    "    print (\"=== Searching query #%d ===\" % qcount)\n",
    "    \n",
    "    with open('12S.list', 'r') as f:\n",
    "        for line in f.readlines():\n",
    "            if fullquery in line:\n",
    "                fullcount += 1\n",
    "                if line not in full_seen:\n",
    "                     full_hit.write(line)\n",
    "                full_seen.add(line)\n",
    "            if squery in line:\n",
    "                scount += 1\n",
    "                if line not in species_seen:\n",
    "                     species_hit.write(line)\n",
    "                species_seen.add(line)\n",
    "            if gquery in line:\n",
    "                gcount += 1 \n",
    "                if line not in genus_seen:\n",
    "                     genus_hit.write(line)\n",
    "                genus_seen.add(line)\n",
    "\n",
    "    print(\"%s hits (full taxonomy): %d\" % (fullquery,fullcount))\n",
    "    print(\"%s hits (species): %d\" % (squery,scount))\n",
    "    print(\"%s hits (genus): %d\" % (gquery,gcount))\n",
    "    \n",
    "    i += 1\n",
    "\n",
    "full_hit.close()\n",
    "species_hit.close()\n",
    "genus_hit.close()\n",
    "\n",
    "print (\"==== Run complete! ===\")\n",
    "print (\"Accession number of full taxonomy hits and description saved in <%s_fulltaxonomy.hits.txt>\" % output_prefix)\n",
    "print (\"Accession number of species hits and description saved in <%s_species.hits.txt>\" % output_prefix)\n",
    "print (\"Accession number of genus hits and description saved in <%s_genus.hits.txt>\" % output_prefix)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
